python/lsst/ts/salkafka/component_producer_set.py

```diff
@@ -116,17 +176,113 @@ def signal_handler(self):
+            # task to wait until process is terminated
+            wait_forever_task = asyncio.Future()
+
+            # Add handle for process termination signals
+            for s in (signal.SIGTERM, signal.SIGINT, signal.SIGHUP):
+                loop.add_signal_handler(s, wait_forever_task.cancel)
+
+            with concurrent.futures.ProcessPoolExecutor(
+                max_workers=len(component_info["topic_sets"]) + 1
+            ) as pool:
+                producer_set = []
+                for topic_set in component_info["topic_sets"]:
+                    producer_set.append(
+                        loop.run_in_executor(
+                            pool,
+                            functools.partial(
+                                cls.create_process,
+                                args=args,
+                                queue_len=component_info.get(
+                                    "queue_len", salobj.topics.DEFAULT_QUEUE_LEN
+                                ),
+                                component=component_info["component"],
+                                add_ack=topic_set.get("add_ack", False),
+                                commands=topic_set.get("commands", []),
+                                events=topic_set.get("events", []),
+                                telemetry=topic_set.get("telemetry", []),
+                            ),
+                        )
+                    )
+
+                # In case one of the parallel producers fails, need to cancel
+                # all the others and exit. This will process any task that
+                # finished and then, proceed to cancel the remaining.
+                main_tasks = [wait_forever_task] + producer_set
+
+                for completed in asyncio.as_completed(main_tasks):
+                    try:
+                        await completed
+                    except asyncio.CancelledError:
+                        print("Terminating process.")
+                    except Exception:
+                        traceback.print_exc()
+                    finally:
+                        break
+
+                # If we get here, it means that it either received a term
+                # signal or one of the child processes failed.
+                # Now it need to send each remaining child process a TERM
+                # signal as well, before exiting.
+                # Gather information about remaining child processes and send
+                # signal.
+
+                # This is the parent process id.
+                parent = psutil.Process(os.getpid())
+
+                # Child process ids.
+                children = parent.children(recursive=True)
+
+                # Send SIGTERM regardless of what signal was received. This
+                # is the safest signal.
+                for process in children:
+                    print(f"Killing child process {process}.")
+                    process.send_signal(signal.SIGTERM)
```

Comment:
I think you don't need all this. I read https://stackoverflow.com/questions/63729195/how-to-terminate-loop-run-in-executor-with-processpoolexecutor-gracefully and played around a bit. SIGTERM is sent automatically through to the child processes and they will exit gracefully without any need for psutil.

Reply:
As I mention in slack, and is also in the link above: 

```
If you are generating the SIGINT with Ctrl+C, then the signal is being sent to the entire process group.
```

which is not the case when we run on k8s.

Reply:
Let's talk about this. It's certainly acceptable if you need it, but I'm wondering if k8s can be configured to send SIGTERM in such a way as to render it unnecessary (the equivalent of `kill -TERM -pid` where pid is the process ID).